ibis/common/graph.py

```diff
@@ -56,19 +56,18 @@ def _flatten_collections(node: Any) -> Iterator[N]:
     >>> c = MyNode(2, "c", (a, b))
     >>> d = MyNode(1, "d", (c,))
     >>>
-    >>> assert list(_flatten_collections(a)) == [a]
     >>> assert list(_flatten_collections((c,))) == [c]
     >>> assert list(_flatten_collections([a, b, (c, a)])) == [a, b, c, a]
+    >>> assert list(_flatten_collections([{"b": b, "a": a}])) == [b, a]
 
     """
-    if isinstance(node, Node):
-        yield node
-    elif isinstance(node, (tuple, list)):
-        for item in node:
+    for item in node:
```

Comment:
Seems like we should perhaps bite the bullet and convert this to an iterative algorithm and remove recursion entirely instead of investing effort in smaller improvements.

Reply:
I can put up a follow up.

Reply:
Seems within noise level to me. Both approaches are an improvement, and I definitely prefer the iterative version for two reasons:

1. There's effectively no contest for deeply nested collections, iteration will win every time in a language that doesn't implement tail recursion
2. The iterative approach uses far less memory, because it's not storing entire stack frames for every nesting level.

Reply:
I'll repost this comment in #8713